### ISSUE TITLE
[JENKINS-50191] Multibranch Pipeline tends to keep rebuilding same commits

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#22643]
<p>We have a number of Pipeline jobs set up with automatic generation through Organization folders (both Stash and Github) and MultiBranch Pipelines generated from those (and finally PR/branch leaf jobs wherever Jenkinsfile's are found). Our default pipeline template includes an SCM polling setup (ranging from H/2 to H/15) because we don't have webhook ability due to corporate firewalling.</p>

<p>Quite often I see builds of the same commit IDs running in parallel, or even the later build starting hours after an earlier one has finished (even if successfully). They often have different causes, such as "Started by an SCM change" vs. "Branch indexing".</p>

<p>If a branch or PR has errors building (e.g. bad timing - new commit noticed in the original branch for a PR, but PR merged or closed during the half a minute or so while a build is preparing, so code can not be checked out - but MBP re-scanning did not yet happen to disable the job), then we can have hundreds of such failed rebuild attempts.</p>

<p>During FOSDEM-2018 discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> (I'm not sure if we logged a ticket at that time) we concluded that such behavior is a bug, and that Jenkins is supposed to track in some matter which source commits it had processed (or even began processing - has queued or is building) and should avoid such automated duplicate re-runs of the same workload, regardless of whether the workspace (on an agent or master) is persistent or removed after each build.</p>

<p>This does not happen often (e.g. every time), but sufficiently frequently to be noticeable and annoying (and sometimes expensive - e.g. if this kicks off the avalanche of product build and re-testing that costs several hours of blocked resources here and there).</p>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jimklimov">jimklimov</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191" target="_blank">Multibranch Pipeline tends to keep rebuilding same commits</a></i></summary>
<i><ul>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>component(s)</b>: core, git-plugin, scm-api-plugin
<li><b>resolution</b>: Unresolved
<li><b>votes</b>: 7
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We have a number of Pipeline jobs set up with automatic generation through Organization folders (both Stash and Github) and MultiBranch Pipelines generated from those (and finally PR/branch leaf jobs wherever Jenkinsfile's are found). Our default pipeline template includes an SCM polling setup (ranging from H/2 to H/15) because we don't have webhook ability due to corporate firewalling.</p>

<p>Quite often I see builds of the same commit IDs running in parallel, or even the later build starting hours after an earlier one has finished (even if successfully). They often have different causes, such as "Started by an SCM change" vs. "Branch indexing".</p>

<p>If a branch or PR has errors building (e.g. bad timing - new commit noticed in the original branch for a PR, but PR merged or closed during the half a minute or so while a build is preparing, so code can not be checked out - but MBP re-scanning did not yet happen to disable the job), then we can have hundreds of such failed rebuild attempts.</p>

<p>During FOSDEM-2018 discussion with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=markewaite" class="user-hover" rel="markewaite">markewaite</a> (I'm not sure if we logged a ticket at that time) we concluded that such behavior is a bug, and that Jenkins is supposed to track in some matter which source commits it had processed (or even began processing - has queued or is building) and should avoid such automated duplicate re-runs of the same workload, regardless of whether the workspace (on an agent or master) is persistent or removed after each build.</p>

<p>This does not happen often (e.g. every time), but sufficiently frequently to be noticeable and annoying (and sometimes expensive - e.g. if this kicks off the avalanche of product build and re-testing that costs several hours of blocked resources here and there).</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/danielbeck.png" title="danielbeck's avatar" /> danielbeck:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=332299&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-332299">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Checkout needs to have finished for a build to be able to make subsequent SCM operations consider the commit already handled. Is this the problem? Huge checkouts, or perhaps even checkouts that aren't happening at the start of the build?</p>
</pre>
</details>

</ul>
</details>

<p>Checkout needs to have finished for a build to be able to make subsequent SCM operations consider the commit already handled. Is this the problem? Huge checkouts, or perhaps even checkouts that aren't happening at the start of the build?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/apgray.png" title="apgray's avatar" /> apgray:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=351472&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-351472">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Seeing the same behaviour on Windows and latest LTS release 2.138.2 (however have seen the same behaviour since at least 2.107.x).</p>

<p>Our situation:</p>

<p>Multibranch pipeline (via a pollSCM(* * * * *) in the Jenkinsfile) will detect an SCM change (size of commit is irrelevant) and kick off a build.  Within a minute another build will be kicked off under the auspices of "Branch Indexing" (via the "Scan Multibranch Pipeline Triggers" setting).</p>

<p>This second duplicate "Branch Indexing build" is buggy behaviour.  The pollSCM renders the branch Indexing build redundant.  Messy workaround is to manually cancel second duplicate build.</p>
</pre>
</details>

</ul>
</details>

<p>Seeing the same behaviour on Windows and latest LTS release 2.138.2 (however have seen the same behaviour since at least 2.107.x).</p>

<p>Our situation:</p>

<p>Multibranch pipeline (via a pollSCM(* * * * *) in the Jenkinsfile) will detect an SCM change (size of commit is irrelevant) and kick off a build.  Within a minute another build will be kicked off under the auspices of "Branch Indexing" (via the "Scan Multibranch Pipeline Triggers" setting).</p>

<p>This second duplicate "Branch Indexing build" is buggy behaviour.  The pollSCM renders the branch Indexing build redundant.  Messy workaround is to manually cancel second duplicate build.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nikhilnanal.svg" title="nikhilnanal's avatar" /> nikhilnanal:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=379548&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379548">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am facing the same issue with a multibranch pipeline (Jenkins 2.164, git hub branch source plugin 2.5.3, git plugin 3.10.0, scm -api 2.4.1). I am using scm polling enabled via the job configuration. If a PR build fails then it gets restarted arbitrarily during a subsequent polling interval even though there was no new change or pull request opened. These restarted PR builds hog my executors and if  a subsequent new PR comes in it has to wait  for the existing builds to complete or I have to manually cancel them. This is a serious problem for us. </p>

<p>I am using "Merging the pull request with current target branch revision" option in my builds. could this be happening if a PR failed but other PRs passed which caused the target revision to be updated and hence the failed PR also gets retriggered?</p>
</pre>
</details>

</ul>
</details>

<p>I am facing the same issue with a multibranch pipeline (Jenkins 2.164, git hub branch source plugin 2.5.3, git plugin 3.10.0, scm -api 2.4.1). I am using scm polling enabled via the job configuration. If a PR build fails then it gets restarted arbitrarily during a subsequent polling interval even though there was no new change or pull request opened. These restarted PR builds hog my executors and if  a subsequent new PR comes in it has to wait  for the existing builds to complete or I have to manually cancel them. This is a serious problem for us. </p>

<p>I am using "Merging the pull request with current target branch revision" option in my builds. could this be happening if a PR failed but other PRs passed which caused the target revision to be updated and hence the failed PR also gets retriggered?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jimklimov.png" title="jimklimov's avatar" /> jimklimov:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=379655&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-379655">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>For the bit about PR-merge builds, yes this is expected to re-run (once per seen target branch update though!) and part of the design idea about such ephemeral merge branches.</p>
</pre>
</details>

</ul>
</details>

<p>For the bit about PR-merge builds, yes this is expected to re-run (once per seen target branch update though!) and part of the design idea about such ephemeral merge branches.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jaygeorgewtc:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=395555&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-395555">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>we are also facing similar issues. where an old build is kicked off of a stale branch via branch indexing where no changes is made with a same commit. which means the pipeline is deploying old code which causes huge problems for the dev team.  which is close to what other people are experiencing. </p>

<p>Our versions </p>

<p>Jenkins 2.222.3</p>

<p>git hub branch source plugin  2.6.0</p>

<p>git hub api  1.111</p>

<p>scm-api  2.6.3</p>
</pre>
</details>

</ul>
</details>

<p>we are also facing similar issues. where an old build is kicked off of a stale branch via branch indexing where no changes is made with a same commit. which means the pipeline is deploying old code which causes huge problems for the dev team.  which is close to what other people are experiencing. </p>

<p>Our versions </p>

<p>Jenkins 2.222.3</p>

<p>git hub branch source plugin  2.6.0</p>

<p>git hub api  1.111</p>

<p>scm-api  2.6.3</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/allan_burdajewicz.png" title="allan_burdajewicz's avatar" /> allan_burdajewicz:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-50191?focusedId=450511&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-450511">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p> SCM Polling is based on the polling baseline / last built revision. Multibranch Pipeline as far as I know is based on the the last recorded / seen revision (the <tt>scm-revision-hash.xml</tt> and recently the <tt>scm-last-seen-revision-hash.xml</tt>: <a href="https://github.com/jenkinsci/branch-api-plugin/blob/2.1200.v4b_a_3da_2eb_db_4/src/main/java/jenkins/branch/BranchProjectFactory.java#L154-L210" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/branch-api-plugin/blob/2.1200.v4b_a_3da_2eb_db_4/src/main/java/jenkins/branch/BranchProjectFactory.java#L154-L210</a>. I am not sure if SCM Polling integrates well with the Multibranch APIs.</p>

<p>Note that a similar problem is observed with just SCM API / Branch API mechanism, see <a href="https://issues.jenkins.io/browse/JENKINS-74839" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-74839</a>. But since there is no SSCM Polling involved I created a different one.</p>
</pre>
</details>

</ul>
</details>

<p> SCM Polling is based on the polling baseline / last built revision. Multibranch Pipeline as far as I know is based on the the last recorded / seen revision (the <tt>scm-revision-hash.xml</tt> and recently the <tt>scm-last-seen-revision-hash.xml</tt>: <a href="https://github.com/jenkinsci/branch-api-plugin/blob/2.1200.v4b_a_3da_2eb_db_4/src/main/java/jenkins/branch/BranchProjectFactory.java#L154-L210" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/branch-api-plugin/blob/2.1200.v4b_a_3da_2eb_db_4/src/main/java/jenkins/branch/BranchProjectFactory.java#L154-L210</a>. I am not sure if SCM Polling integrates well with the Multibranch APIs.</p>

<p>Note that a similar problem is observed with just SCM API / Branch API mechanism, see <a href="https://issues.jenkins.io/browse/JENKINS-74839" class="external-link" rel="nofollow">https://issues.jenkins.io/browse/JENKINS-74839</a>. But since there is no SSCM Polling involved I created a different one.</p>
